src/material/tabs/tab-nav-bar/tab-nav-bar.ts

```diff
@@ -274,7 +293,31 @@ export class _MatTabLinkBase
   _handleFocus() {
     // Since we allow navigation through tabbing in the nav bar, we
     // have to update the focused index whenever the link receives focus.
-    this._tabNavBar.focusIndex = this._tabNavBar._items.toArray().indexOf(this);
+    this._tabNavBar.focusIndex = this._getIndex();
+  }
+
+  _handleSpace() {
+    if (this._tabNavBar.panel) {
+      this.elementRef.nativeElement.click();
+    }
+  }
+
+  /** Returns the tab's tabindex. */
+  _getTabIndex(): number {
+    if (!this._tabNavBar.panel) {
+      return this.tabIndex;
+    }
+
+    if (!this._tabNavBar._items) {
+      return this._isActive ? 0 : -1;
+    }
+
+    return this._tabNavBar.focusIndex === this._getIndex() ? 0 : -1;
```

Comment:
Do we really need to change the tabindex based on the focus index? If focus is already on the tab, the tabindex doesn't matter anymore.

Reply:
Currently, `MatTabNav` has by default `tabindex="0"` on all tabs. This means that every tab is in the tab order, which is incorrect for managed focus reasons (per the ARIA spec). Thus, this logic implements the [roving tabindex pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex).

Note that `MatTabGroup`'s roving tabindex implementation is actually slightly incorrect, this one I added to `MatTabNav` _is_ correct. It's important for the `tabindex="0"` to track focus, as opposed to selection. This way if you tab to a `MatTabNav`, arrow among some tabs, tab out, and tab back in, the same tab you were previously focused on is still focused.

Reply:
Okay, that makes sense, but I think that `_getIndex` might have to be optimized. Since it's used in a binding, it'll be called on each change detection and it currently uses a linear search to figure out the index. Maybe it can be saved in a property and updated when the list of tabs changes?

Reply:
Done.

Reply:
Done.

Reply:
I ended up reverting the roving tabindex behavior because it's actually tricky to implement with `FocusKeyManager` without triggering ECAC errors. The problem is that the tabindex needs to be changed on focus, and it's common for people to move focus in `ngAfterViewInit` lifecycles as well as other problematic contexts.

Note that I actually ran into this a year or so ago with `MatSelectionList`. From https://github.com/angular/components/pull/18445:

> I would prefer to use the "roving tabindex" pattern, but given the list options are content-projected, I ran into issues keeping tabindices updated without ExpressionChangedAfter... errors. Thus, I instead essentially followed the implementation used in MatChipList. The listbox element has tabindex="0", and the option elements have tabindex="-1". When the listbox is focused, it redirects focus to the proper option element.

Instead of doing the same thing here, I decided to just match `MatTabGroup`'s behavior and have the tabindex rove on selection as opposed to focus. The reason I decided to do this was because this PR is large as-is. We can always improve the roving tabindex behavior in both tabs components in subsequent PRs.